fem/hyperbolic_conservation_laws.hpp

```diff
@@ -0,0 +1,779 @@
+//           MFEM Hyperbolic Conservation Laws Serial/Parallel Shared Class
+//
+// Description:  This file contains general hyperbolic conservation element/face
+//               form integrators.
+//
+//               Abstract RiemannSolver, HyperbolicFormIntegrator, and
+//               DGHyperbolicConservationLaws are defined. Also, several example
+//               hyperbolic equations are defined; including advection, burgers,
+//               shallow water, and euler equations.
+//
+//               To implement a specific hyperbolic conservation laws, users can
+//               create derived classes from HyperbolicFormIntegrator with overloaded
+//               ComputeFlux. One can optionally overload ComputeFluxDotN to avoid
+//               creating dense matrix when computing normal flux.
+//
+//               FormIntegrator use high-order quadrature points to implement
+//               the given form to handle nonlinear flux function. During the
+//               implementation of forms, it updates maximum characteristic
+//               speed and collected by DGHyperbolicConservationLaws. The global
+//               maximum characteristic speed can be obtained by public method
+//               getMaxCharSpeed so that the time step can be determined by CFL
+//               condition. Also, resetMaxCharSpeed should be called after each Mult.
+//
+//               @note For parallel version, users should reduce all maximum
+//               characteristic speed from all processes using MPI_Allreduce.
+//
+// @todo: Implement limiter @dohyun-cse
+// @todo: Implement boundary condition @dohyun-cse
+//
+//
+// Class structure: DGHyperbolicConservationLaws
+//                  |- HyperbolicFormIntegrator: (F(u,x), grad v) and <F̂(u, x, n), [[v]])
+//                  |  |- RiemannSolver: (F±, u±) ↦ F̂(u,x,n)
+//                  |
+//                  |- (Par)NonlinearForm: Evaluate form integrators
+//
+
+#ifndef MFEM_HYPERBOLIC_CONSERVATION_LAWS
+#define MFEM_HYPERBOLIC_CONSERVATION_LAWS
+
+#include "nonlinearform.hpp"
+#ifdef MFEM_USE_MPI
+#include "pnonlinearform.hpp"
+#endif
+namespace mfem
+{
+
+/**
+ * @brief Abstract class for numerical flux for an hyperbolic conservation laws
+ * on a face with states, fluxes and characteristic speed
+ *
+ */
+class RiemannSolver
+{
+public:
+   RiemannSolver() {}
+   /**
+    * @brief Evaluates numerical flux for given states and fluxes. Must be
+    * overloaded in a derived class
+    *
+    * @param[in] state1 state value at a point from the first element
+    * (num_equations x 1)
+    * @param[in] state2 state value at a point from the second element
+    * (num_equations x 1)
+    * @param[in] fluxN1 normal flux value at a point from the first element
+    * (num_equations x dim)
+    * @param[in] fluxN2 normal flux value at a point from the second element
+    * (num_equations x dim)
+    * @param[in] maxE maximum characteristic speed (eigenvalue of flux jacobian)
```

Comment:
Not all Riemann solvers need to take the max characteristics speed as input (and anyway, this can be computed from the two incoming fluxes).

Also, in general it will not be equal to the maximal eigenvalue of the Jacobian

Reply:
I removed eigenvalue of Jacobian from the comment.
But I want to keep the characteristic speed (I changed it from maxE to speed1, speed2) as input.
Characteristic speed (char speed) is one of the most common parameters for the approximated Riemann solver. While it can be computed from the fluxes, it is equation-specific. Also, it is already computed from the flux evaluation. Therefore, there is no need to re-compute this.

Reply:
I think it's more consistent with the mathematical definition if `RiemannSolver::Eval` takes as inputs two states, and a normal vector. That is enough information to compute the solution to the Riemann problem, but it means that the `RiemannSolver` object needs to also know the governing equations.

Perhaps it makes sense to have the equations represented as an object (separate from the integrator) that provides `ComputeFlux` and `ComputeFluxDotN`. Both `RiemannSolver` and `HyperbolicFormIntegrator` would take as input this equation object. Specific subclasses of `HyperbolicFormIntegrator` would not be required.

Reply:
Thank you for the suggestion! Now I refactored the `hyperbolic_conservation_laws.hpp` so that the form integrator and Riemann solver takes `FluxFunction` as input. Check out the new implementation :)